src/hangar/dataset/numpy_dset.py

```diff
@@ -0,0 +1,238 @@
+from typing import Sequence, Callable, TYPE_CHECKING, Union, List, Tuple
+import random
+
+import numpy as np
+
+from .common import HangarDataset
+
+if TYPE_CHECKING:
+    from ..columns import ModifierTypes
+    Columns = ModifierTypes
+    KeyType = Union[str, int, List, Tuple]
+
+
+def default_collate_fn(col_functions):
+    def wrapper(data_arr):
+        # data_arr -> array of data samples from each column
+        collated = []
+        for fn in col_functions:
+            grouped = fn(data_arr)
+            collated.append(grouped)
+        return tuple(collated)
+    return wrapper
+
+
+class NumpyDataset:
+    """NumpyDataset class provides interfaces for users to iterate over the batches of
+    data from different columns. The only user facing APIs it exposes are ``__len__`` and
+    ``__iter__``. Batch and shuffle operations are handled by `:func:`make_numpy_dataset`
+    based on the arguments it gets and hence user should not interact with this class for
+    such operations. Note that, user would never instantiate this class directly. Instead
+    `:func:`make_numpy_dataset` act as the entry point and return an object of this class
+    to the user
+
+    Parameters
+    ----------
+    dataset : :class:`~hangar.dataset.common.Dataset` object
+        Hangar's Dataset object that groups columns for downstream processing
+    batch_size : int
+        Size of the individual batch. If specified batches of this size will be returned
+        on each iteration
+    drop_last : bool
+        Should drop the last incomplete batch
+    shuffle : bool
+        Should shuffle the batch on each epoch
+    collate_fn : Callable
+        A function to collate samples together in a batch. In case this option is absent,
+        the heuristics to collate the batch is
+            1. If the column is an ndarray flat column, then `np.stack` will be used
+            2. If the column is with any other properties, `list.append` will be used
+        Note that the batch of data that comes to callate_fn will have each elements consist
+        of datapoints from all the columns. For example, if the columns from where the data
+        being fetched are col1 and col2 then the batch would look like
+
+        ```python
+        [
+            (data0_col1, data0_col2),
+            (data1_col1, data1_col2),
+            (data2_col1, data2_col2),
+            ...
+        ]
+        ```
+    """
+    def __init__(self, dataset: HangarDataset, batch_size: int, drop_last: bool,
+                 shuffle: bool, collate_fn: Callable = None):
+        self._dataset = dataset
+        self._num_batches = None
+        self._batch_size = None
+        if batch_size:
+            if not collate_fn:
+                collate_colfn = []
+                for col in dataset.columns.values():
+                    print(col.column_type)
```

Comment:
@rlizzo We might need to remove this print?

Reply:
Haha yeah I think that's one left over from Debugging.